### PR TITLE
Added config options for junimo unlocks, locations, & bait maker

### PR DESCRIPTION
**Pull request by BussinBungus**

### CODE REVIEW

**Comment by NermNermNerm** [comment]
I'm always a bit down on settings.  It feels the use of settings is a thing you do when you don't really understand your userbase.  I'm not saying I won't end up taking this mod as-is, but let's think it through a bit.

There are the use-cases that I know for sure that are currently being ignored that would be addressed by this PR.  I'm breaking it down by group.

For the quest-bypass overrides, it helps these use-cases:
1. Folks just want to test out the mod sometimes.  They want to see what they're getting if they go to the trouble to unlock the things.
2. Other mods make doing the quests impractical or impossible.
3. Some people get mods to win faster.

The use-case for the location-restriction easing are:

1. Some people get mods to win faster.
2. Other mods might provide locations that are compatible with the spirit of the restriction in this mod.

I really have no idea why hooking up the bait machine would ever be a good idea.  Are there mods that would make having bulk-species-specific-bait a good idea?  Seems to me more likely you'll catch the Legend and drop him in the mix hoping to get smoked Legend worth a kabillion gold and get rewarded with 6 bits of worthless bait because, weeks ago, the player fiddled with the setting because "OF COURSE I WANT THIS MACHINE TURNED ON!  I NEED ALL THE MACHINES!  ALL THE MACHINES!"  and then the legend mishap happens and we get a bug: "YOUR MOD IS STUPID!  IT SHOULD ALWYAS MOVE THE FISH TO THE SMOKER IF IT CNA!"

The "I just want to win faster" kind of user would be better served with Automate.  But perhaps it's the draw of the animation that brings them in -- maybe they don't care about the game-balance problems of Automate and just want the animation that the mod brings.  For that kind of user, wouldn't a single "Unlock Everything" setting be sufficient?  That would also probably suit the needs of the player who just wants to try it out too.  The only user that this leaves out would be the ones that would love to do all the quests, but can't due to some incompatibility or another.

Another question to ask is whether this is a setting or a cheat-code?  If it's a cheat-code, it could be done a whole lot simpler with an SMAPI command.

For the location setting, I just about think it's a necessity.  There's other ways I'd like to implement it besides a hard-coded list, but I'd have to agree that we need a way for users to take control over it because there's an infinite variety of mods.


I hope I don't sound like I'm totally bagging on all your work!  I'm just trying to offer a different point of view.  Feel free to point out where I'm wrong!

File: Junimatic/ModConfigMenu.cs [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+                save: () => this.mod.Helper.WriteConfig(Config)
+            );
+
+            configMenu.AddSectionTitle(
+                mod: ModManifest,
+                text: () => "Unlocks"
+            );
+            configMenu.AddBoolOption(
+                mod: ModManifest,
+                name: () => "Unlock Junimo Portal",
```
Comment (NermNermNerm) [inline Junimatic/ModConfigMenu.cs:46]: I wonder if users would be confused by this.  E.g. might they click on "Unlock Animal Junimos" below and not this, then wonder why they can't build a network?

File: Junimatic/WorkFinder.cs [hunk 2]
```diff
@@ -111,16 +111,23 @@ private void DoJunimos(bool isAutomationInterval)
...
                     .Where(l => l is not null).Select(l => l!)
                     .ToList();
-            allJunimoFriendlyLocations.Add(Game1.getFarm());
-            allJunimoFriendlyLocations.AddRange(
-                new string[] { "FarmCave", "IslandWest", "Cellar", "FarmHouse", "IslandFarmHouse" }
-                .Select(name => Game1.getLocationFromName(name))
-                .Where(l => l is not null));
+                allJunimoFriendlyLocations.Add(Game1.getFarm());
+                allJunimoFriendlyLocations.AddRange(
+                    new string[] { "FarmCave", "IslandWest", "Cellar", "FarmHouse", "IslandFarmHouse" }
```
Comment (NermNermNerm) [inline Junimatic/WorkFinder.cs:127]: Perhaps instead of having the setting Extend this list - maybe this list should be the default value of the setting?

**Comment by BussinBungus** [comment]
Your review raises a lot of good points. I personally like mods with many settings so I can tweak them how I like, but I can also see how that approach can end with a cluttered, confusing config menu.

Starting with the bait maker, I get that most people don't want to accidentally bait their legendary fish. However, I think it should at least be mentioned on the mod page that the bait maker is intentionally disabled so you don't have the opposite problem of people reporting "Bait maker not working" bugs. Also, I'd argue it's the player's fault if they somehow include a bait maker in a network since, unlike Automate, you have to intentionally hook machines up with paths. I could see the bait maker being useful to automate if you have a small network of only bait makers and a chest containing small quantities of many types of fish, so you have bait of any given type on hand when a quest requires a specific fish.

For individual quest-bypass overrides, the main use-case that I can think of is a player wanting to complete most quests, but needing to skip a specific one they find too challenging (or impossible due to other mods). While adding the overrides, I thought another use-case might be players who want to disable automation for a specific category of machine (e.g., crop processors) due to thinking it's too unbalanced, though in hindsight the config wouldn't completely prevent automation of those machines since you could still just complete the related quest. Both cases are admittedly very niche, and I agree an "unlock everything" option would be better for most users. This is essentially a cheat code and could be done via console command, I just don't trust people to be able to use console and thought a GUI would be easier.

I don't have much to add about the location restrictions. I agree with your comment about the default value of the list, I'm just pretty new to programming and C# so I didn't realize that was a better way to do it. I thought about including a text box config field where users can type a comma-separated list of locations to enable if they wanted something a bit more balanced than simply enabling all locations, but that would require players to know the internal names for locations and thus be prone to error.

Another idea I'd like to mention is a "machine blacklist" of sorts, where users can name machines that shouldn't be allowed in networks (if the vanilla code permits this). It might include the shipping bin and bait maker by default, but users could either remove those from the list to enable them or add machines that they feel would be too overpowered to automate (e.g., Casks).

File: Junimatic/ModConfigMenu.cs (hunk 1 above)
Comment (BussinBungus) [reply Junimatic/ModConfigMenu.cs:46]: I agree, it's probably best to combine them all into an "Unlock Everything" command or config.

File: Junimatic/WorkFinder.cs (hunk 2 above)
Comment (BussinBungus) [reply Junimatic/WorkFinder.cs:127]: That's definitely a cleaner way to do it, I just didn't think of that.

**Comment by NermNermNerm** [comment]
I still don't buy the bait-maker use-case.  If you need fish for a quest, the thing to do is to keep the fish for the future quest!  But in any case, the "Machine Blacklist" idea is really good, and should be the way the baitmaker's default ban should be done.

One thing that worries me a bit about the "Machine Blacklist" and "Allowed Locations" is that there's no way in the mod config menu to check the spelling/capitalization/spacing of names and so it's really easy for the user to goof something there and then wonder why it doesn't work like they think it should.  Perhaps we should have code that runs on-day-start and does a LogErrorOnce if there are any locations/machines that don't resolve.

After sleeping on it, I agree that a setting is more discoverable than a console command.  I do wonder whether turning the setting on should disable the quests.  If it does disable the quests, what if they're in-flight?  Would it be confusing to the user to have this quest happen?  If you were just reviewing the mod, then having the quests available would be a good thing, I suppose.

**Comment by nephatrine** [comment]
Downloaded @BussinBungus 's fork here so I could manually enable the fishing junimo since that quest didn't seem to properly work for me and I'm getting tons of these errors in my SMAPI log (though the junimos seem to be functioning and is even smoking my fish now). I wasn't getting them before with the build on the nexus so wanted to point it out here in case it is something funny with the location configuration or something from this PR.

```
[Junimatic] This mod failed in the GameLoop.OneSecondUpdateTicked event. Technical details:
NullReferenceException: Object reference not set to an instance of an object.
   at NermNermNerm.Junimatic.ObjectMachine.IsCompatibleWithJunimoNoCache(JunimoType projectType) in C:\Users\Nephatrine\AppData\Local\ModOrganizer\Stardew Valley\SDV-NermNermNerm-Junimatic-main\Junimatic\ObjectMachine.cs:line 186
   at NermNermNerm.Junimatic.ObjectMachine.IsCompatibleWithJunimo(JunimoType projectType) in C:\Users\Nephatrine\AppData\Local\ModOrganizer\Stardew Valley\SDV-NermNermNerm-Junimatic-main\Junimatic\ObjectMachine.cs:line 110
   at NermNermNerm.Junimatic.WorkFinder.BuildNetwork(GameLocation location) in C:\Users\Nephatrine\AppData\Local\ModOrganizer\Stardew Valley\SDV-NermNermNerm-Junimatic-main\Junimatic\WorkFinder.cs:line 311
   at NermNermNerm.Junimatic.WorkFinder.TryDoAutomationsForLocation(GameLocation location, JunimoType projectType) in C:\Users\Nephatrine\AppData\Local\ModOrganizer\Stardew Valley\SDV-NermNermNerm-Junimatic-main\Junimatic\WorkFinder.cs:line 209
   at NermNermNerm.Junimatic.WorkFinder.DoJunimos(Boolean isAutomationInterval) in C:\Users\Nephatrine\AppData\Local\ModOrganizer\Stardew Valley\SDV-NermNermNerm-Junimatic-main\Junimatic\WorkFinder.cs:line 182
   at NermNermNerm.Junimatic.WorkFinder.GameLoop_OneSecondUpdateTicked(Object sender, OneSecondUpdateTickedEventArgs e) in C:\Users\Nephatrine\AppData\Local\ModOrganizer\Stardew Valley\SDV-NermNermNerm-Junimatic-main\Junimatic\WorkFinder.cs:line 99
   at StardewModdingAPI.Framework.Events.ManagedEvent`1.Raise(TEventArgs args) in SMAPI\Framework\Events\ManagedEvent.cs:line 101
```

**Comment by BussinBungus** [comment]
> One thing that worries me a bit about the "Machine Blacklist" and "Allowed Locations" is that there's no way in the mod config menu to check the spelling/capitalization/spacing of names and so it's really easy for the user to goof something there and then wonder why it doesn't work like they think it should. Perhaps we should have code that runs on-day-start and does a LogErrorOnce if there are any locations/machines that don't resolve.

I think the logging is a good idea. I looked at Automate's config out of curiosity and saw it has a similar "Machine Overrides" system where you can enable/disable and set the priority of individual machines, but idk how difficult that'd be to adapt that vs. making a blacklist system from scratch.
> After sleeping on it, I agree that a setting is more discoverable than a console command. I do wonder whether turning the setting on should disable the quests. If it does disable the quests, what if they're in-flight? Would it be confusing to the user to have this quest happen? If you were just reviewing the mod, then having the quests available would be a good thing, I suppose.

It's up to you of course, but my opinion is that quests should probably be disabled and removed or completed if in-flight when the setting is turned on. I agree that doing them and seeing the unlock events when you've already been using the Junimos would be confusing.
> Downloaded @BussinBungus 's fork here so I could manually enable the fishing junimo since that quest didn't seem to properly work for me and I'm getting tons of these errors in my SMAPI log (though the junimos seem to be functioning and is even smoking my fish now). I wasn't getting them before with the build on the nexus so wanted to point it out here in case it is something funny with the location configuration or something from this PR.

@nephatrine I'm not sure why it's giving that error, hopefully @NermNermNerm can figure it out. It looks like the relevant code is below, where itemData is the object reference without an instance.

    {
        var itemData = ItemRegistry.GetData(trigger.RequiredItemId);
        if (categories[(int)projectType].Contains(itemData.Category)) // ObjectMachine.cs:line 186
        {
            return true;
        }
    }

**Comment by BussinBungus** [comment]
Hi, I think I was mentioned in one of the bug reports about dehydrators and wanted to make sure you're not waiting on me for anything. I only made the settings as a personal edit that I figured I'd share in case it helps. I don't really have plans to change my commit or add more code at the moment, but you're free to use, not use, and/or change as much of it as you'd like! Sorry if there was any confusion.

**Comment by NermNermNerm** [comment]
Shall we close this PR?  Most of the ideas in here have been integrated into the mod.

**Pull request closed without merging**